File: Medium/0003_Longest_Substring_Without_Repeating_Characters/go/optimal.go
package main

import (
	"math"
)

func lengthOfLongestSubstring(s string) int {
	positions := map[string]int{}
	var maxLength float64 = 0
	start := 0
	end := 0

	// Build the map ahead of time (so values aren't rehashed for map expansion as we go)
	for _, c := range s {
		key := string(c)
		positions[key] = -1
	}

	// Scan the string
	for j, c := range s {
		key := string(c)
		i, _ := positions[key]
		positions[key] = j
		end++

		if i < start {
			// Haven't seen this character, recompute max length
			maxLength = math.Max(maxLength, (float64)(end-start))
		} else {
			// Have seen this character, skip forward
			start = i + 1
		}
	}

	return (int)(maxLength)
}

File: Medium/0003_Longest_Substring_Without_Repeating_Characters/python3/optimal.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """
        Method 1: Sliding Window + Hashing: O(n), O(n)
        """
        seen = {}
        left = 0
        length = 0
        for index in range(len(s)):
            character = s[index]
            if character in seen and seen[character] >= left:
                left = seen[character]+1
            seen[character] = index
            length = max(length, index+1-left)
        return length